File: python/Lec 2 Build-up Logical Thinking (Patterns)/pattern_05.py
# *****
# ****
# ***
# **
# *

#----------------------------------
# BRUTE FORCE APPROACH
#----------------------------------
n = 5
for i in range(n, 0, -1):
    line = ""
    for j in range(i):
        line += "★"  # O(n²) string copying
    print(line)

#----------------------------------
# BETTER APPROACH (String multiplication)
#----------------------------------
n = 5
for i in range(n, 0, -1):
    print("★" * i)

# ----------------------------------
#  BEST APPROACH (List comprehension with join)
# ----------------------------------
n = 5
print('\n'.join(["★" * i for i in range(n, 0, -1)]))

File: cpp/Lec 2 Build-up Logical Thinking (Patterns)/pattern_05.cpp
// *****
// ****
// ***
// **
// *
// //----------------------------------
// // BRUTE FORCE APPROACH (Nested loops)
// //----------------------------------
#include <iostream>
using namespace std;
int main()
{
    int n = 5;
    for (int i = n; i >= 1; i--)
    {
        for (int j = 1; j <= i; j++)
        {
            cout << "★";
        }
        cout << endl;
    }
    return 0;
}

//----------------------------------
// BETTER APPROACH (String constructor)
//----------------------------------
#include <iostream>
#include <string>
using namespace std;
int main()
{
    int n = 5;
    for (int i = n; i >= 1; i--)
    {
        cout << string(i, '*') << endl;
    }
    return 0;
}

// //----------------------------------
// // BEST APPROACH (Single loop, pre-allocated)
// //----------------------------------
#include <iostream>
#include <string>
using namespace std;
int main()
{
    int n = 5;
    string s(n, '*');
    for (int i = 0; i < n; i++)
    {
        cout << s << '\n';
        s.pop_back(); // Remove last character
    }
    return 0;
}